src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,118 @@
+        } else if (index < size) {
+            if (size + 1 > elements.length) {
+                grow();
+            }
+            T[] firstSubArray = Arrays.copyOfRange(elements, 0, index + 1);
+            T[] secondSubArray = Arrays.copyOfRange(elements, index, elements.length - 1);
+            firstSubArray[firstSubArray.length - 1] = value;
+            T[] resultArray = Arrays.copyOf(firstSubArray,
+                    firstSubArray.length + secondSubArray.length);
+            System.arraycopy(secondSubArray, 0, resultArray,
+                    firstSubArray.length, secondSubArray.length);
+            elements = resultArray;
+            size++;
+        } else {
+            elements[index] = value;
+            size++;
+        }
     }
-
+    
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
-
+    
     @Override
     public T get(int index) {
-        return null;
+        checkGetIndex(index);
+        return elements[index];
     }
-
+    
     @Override
     public void set(T value, int index) {
-
+        checkGetIndex(index);
+        elements[index] = value;
     }
-
+    
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T[] firstSubArray = Arrays.copyOfRange(elements, 0, index);
+        T[] secondSubArray = Arrays.copyOfRange(elements, index + 1, elements.length);
+        T[] resultArray = Arrays.copyOf(firstSubArray,
+                firstSubArray.length + secondSubArray.length + 1);
+        System.arraycopy(secondSubArray, 0, resultArray,
+                firstSubArray.length, secondSubArray.length);
+        T oldElement = get(index);
+        elements = resultArray;
+        size--;
+        return oldElement;
     }
-
+    
     @Override
     public T remove(T t) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] == t || (elements[i] != null && elements[i].equals(t))) {
+                T removedElement = get(i);
+                remove(i);
+                return removedElement;
+            }
+        }
+        throw new NoSuchElementException("Element \"" + t + "\" not found.");
     }
-
+    
     @Override
     public int size() {
-        return 0;
+        return size;
     }
-
+    
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+    
+    private void grow() {
+        elements = Arrays.copyOf(elements, (int) (elements.length * 1.5));
+    }
+    
+    private void checkIndex(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+    
+    private void checkGetIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
```

Comment:
```suggestion
    private void checkIndex(int index) {
        if (index >= size || index < 0) {
            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
        }
```
this method should be used in `get()`, `set()` and `remove()` methods